File: mobile/modules/mediapipe-face-landmarker/ios/HybridGlowFaceLandmarker.swift
import Foundation
import NitroModules
import MediaPipeTasksVision

// iOS Nitro HybridObject wrapping MediaPipe's Face Landmarker Tasks Vision
// SDK as a live vision-camera frame processor. Modeled directly on
// react-native-vision-camera-ocr-plus's real, working v5/Nitro pattern
// (HybridTextRecognizer.swift) — same non-blocking "isBusy/lastResult"
// shape, same raw-pointer CVPixelBuffer reconstruction — swapped for
// MediaPipe's async detectAsync/delegate API instead of ML Kit's
// synchronous results(in:).
//
// Output is deliberately reduced to the SAME shape
// react-native-vision-camera-face-detector's own `Face` already has
// (bounds + pitch/roll/yaw) — see GlowFaceLandmarker.nitro.ts's own header
// comment — specifically so SkinScanCamera.tsx's existing gating math
// (positionGate/angleGate thresholds, PITCH_GATE_DEG) needs ZERO changes.
// MediaPipe's real output is 478 3D landmarks per face, not a bounding
// box+angles triple — the reduction happens here, once, at the native
// boundary.
//
// bounds is returned as 0-1 NORMALIZED fractions of the frame, NOT
// screen/preview pixel coordinates — unlike the existing ML Kit path,
// which gets pixel coordinates for free from that package's own `autoMode`
// (native-side rotation/scaling against windowWidth/windowHeight). This
// module doesn't replicate that scaling in Swift — deliberately: getting a
// coordinate transform wrong in blind native code is a silent, hard-to-spot
// bug (numbers that still LOOK plausible), whereas doing that same scaling
// in plain JS (see MediaPipeFaceLandmarkerSensor in ../src/index.ts) is
// something that can actually be reasoned about/checked against the
// existing liveBox math. This is a deliberate scope reduction, not an
// oversight — flagged explicitly in this project's own scope report.
//
// BUILD HISTORY (real, not hypothetical): the first version of this file
// failed a real EAS Xcode build (2026-09-01, build #36) with three distinct,
// confirmed errors, each fixed below with the actual reason, not a second
// guess dressed up as confidence:
//  1. "cannot declare conformance to 'NSObjectProtocol' in Swift;
//     'HybridGlowFaceLandmarker' should inherit 'NSObject' instead" — this
//     class conformed to FaceLandmarkerLiveStreamDelegate directly, but its
//     base (HybridGlowFaceLandmarkerSpec_base, Nitrogen-generated) is a
//     plain Swift class, not NSObject-derived, and MediaPipe's delegate
//     protocol is Objective-C-bridged (needs NSObjectProtocol). Fixed by
//     moving the delegate conformance to a small private NSObject proxy
//     (FaceLandmarkerDelegateProxy below) that forwards to this class
//     instead of conforming directly.
//  2/3. "binary operator '>' cannot be applied to operands of type
//     'CGFloat' and 'Float'" / "cannot assign value of type 'CGFloat' to
//     type 'Float'" — landmark coordinates (NormalizedLandmark.x/.y) are
//     actually CGFloat, not Float as originally assumed; boundingBox and
//     the largest-face comparison below were mixing the two. Fixed by using
//     CGFloat throughout those two spots.
//  The build ALSO reported seven "value of type 'UInt' has no member '2'"
//  errors, all traced to the facialTransformationMatrixes-based pitch/yaw
//  decomposition this file originally had (`m.columns.2.x` etc.) — that
//  code assumed the matrix type was simd_float4x4 with no real confirmation
//  it is, and this error doesn't even look like a normal simd_float4x4
//  access failure. Rather than guess a SECOND time blind and burn another
//  build finding out, that whole decomposition is REMOVED here, not
//  re-guessed — pitch/yaw fall back to 0 (reads as "straight") the same way
//  they already did when the matrix was unavailable. This is a real,
//  deliberate scope cut, not a bug fix — see this project's own
//  verification report for what real ML Kit comparison and a correct
//  pitch/yaw source would take to add back.
class HybridGlowFaceLandmarker: HybridGlowFaceLandmarkerSpec {

  private var landmarker: FaceLandmarker?
  private var delegateProxy: FaceLandmarkerDelegateProxy?
  private let stateLock = NSLock()
  private var isBusy: Bool = false
  private var lastResult: [DetectedFace] = []

  override init() {
    super.init()
    setupLandmarker()
  }

  private func setupLandmarker() {
    // Bundle(for:) resolves against THIS class's own bundle, i.e. the pod's
    // resource bundle the podspec declares (see
    // NitroGlowFaceLandmarker.podspec's resource_bundles) — not the host
    // app's main bundle, which wouldn't have this file.
    guard let modelPath = Bundle(for: HybridGlowFaceLandmarker.self)
      .path(forResource: "face_landmarker", ofType: "task") else {
      // No crash, no throw — same "fails to empty, never brings down the
      // screen" convention LightingSensor's own ErrorBoundary already
      // established for an equivalent "native piece not linked/built yet"
      // case. detectFrame below simply returns [] forever if this happens,
      // which SkinScanCamera.tsx's MediaPipe path already treats as "no
      // face" — see this module's src/index.ts.
      print("[GlowFaceLandmarker] face_landmarker.task not found in bundle — detection will return no faces")
      return
    }
    let options = FaceLandmarkerOptions()
    options.baseOptions.modelAssetPath = modelPath
    options.runningMode = .liveStream
    // This app only ever cares about ONE face (the user's own, live
    // framing) — same as the existing ML Kit path only ever consuming the
    // single largest detected face (see SkinScanCamera.tsx's
    // primaryLiveFace).
    options.numFaces = 1
    // See this class's own header for why: real, but currently unused
    // (facialTransformationMatrixes decomposition was removed after a
    // confirmed build failure) — left set so it's a one-line re-enable
    // once a correct decomposition replaces the removed one, rather than
    // needing to rediscover this option exists.
    options.outputFacialTransformationMatrixes = true
    // See FaceLandmarkerDelegateProxy below — this class itself no longer
    // conforms to FaceLandmarkerLiveStreamDelegate (that's error #1 in this
    // file's own header).
    let proxy = FaceLandmarkerDelegateProxy(owner: self)
    self.delegateProxy = proxy
    options.faceLandmarkerLiveStreamDelegate = proxy
    landmarker = try? FaceLandmarker(options: options)
    if landmarker == nil {
      print("[GlowFaceLandmarker] FaceLandmarker failed to initialize")
    }
  }

  // MARK: - HybridGlowFaceLandmarkerSpec

  func detectFrame(nativeBufferPointer: UInt64, orientation: String) throws -> [DetectedFace] {
    guard let landmarker = landmarker else { return [] }

    stateLock.lock()
    let busy = isBusy
    let current = lastResult
    stateLock.unlock()
    // Never blocks the frame thread on MediaPipe inference — if the
    // previous frame's detectAsync call hasn't completed yet (delegate not
    // fired), skip submitting a new one and hand back the last real result,
    // exactly like the OCR-plus reference's scanFrame does for ML Kit OCR.
    if busy { return current }

    guard let rawPtr = UnsafeRawPointer(bitPattern: UInt(nativeBufferPointer)) else { return current }
    let pixelBuffer = Unmanaged<CVPixelBuffer>.fromOpaque(rawPtr).takeUnretainedValue()

    // MPImage(pixelBuffer:orientation:) — the orientation-aware initializer
    // is the one actually needed here (a live front-camera frame is NOT
    // upright the way a file-loaded UIImage already is); Google's own
    // published docs snippet only shows the bare MPImage(pixelBuffer:)
    // default-orientation form, so this exact overload's existence/exact
    // parameter name is UNVERIFIED against the real MediaPipeTasksVision
    // header — this did NOT show up as one of the confirmed build errors
    // (the build never got this far — it failed at the delegate-conformance
    // error first, before any of detectFrame's own code was reached by the
    // type checker in a way that would surface a problem here), so this
    // is still genuinely untested, not confirmed-working. If this overload
    // doesn't exist, the real fix is almost certainly still an
    // orientation-aware MPImage constructor, just possibly named
    // differently — flagged here plainly rather than presented as certain.
    guard let image = try? MPImage(pixelBuffer: pixelBuffer, orientation: Self.imageOrientation(from: orientation)) else {
      return current
    }

    stateLock.lock()
    isBusy = true
    stateLock.unlock()
    let timestampMs = Int(Date().timeIntervalSince1970 * 1000)
    do {
      try landmarker.detectAsync(image: image, timestampInMilliseconds: timestampMs)
    } catch {
      // Submission itself failed (not a detection failure — those come
      // back through the delegate below) — clear isBusy so the NEXT frame
      // isn't permanently blocked believing one is still in flight.
      stateLock.lock()
      isBusy = false
      stateLock.unlock()
    }
    return current
  }

  // MARK: - Result handling (called by FaceLandmarkerDelegateProxy below)

  fileprivate func handleResult(
    _ result: FaceLandmarkerResult?,
    timestampInMilliseconds: Int,
    error: Error?
  ) {
    defer {
      stateLock.lock()
      isBusy = false
      stateLock.unlock()
    }
    guard let result = result, !result.faceLandmarks.isEmpty else {
      stateLock.lock()
      lastResult = []
      stateLock.unlock()
      return
    }

    // Largest face wins — same convention as detectFaceRegion
    // (SkinScanCamera.tsx) and the existing live ML Kit path
    // (primaryLiveFace) already use, applied here via landmark bounding-box
    // area since MediaPipe doesn't hand back a pre-computed box the way ML
    // Kit's Face.bounds does. CGFloat throughout (NOT Float) — see this
    // file's own header, error #2/#3: NormalizedLandmark.x/.y are CGFloat,
    // confirmed by a real build failure when this mixed the two.
    var biggestIndex = 0
    var biggestArea: CGFloat = -1
    for (i, landmarks) in result.faceLandmarks.enumerated() {
      let box = Self.boundingBox(landmarks)
      let area = box.width * box.height
      if area > biggestArea { biggestArea = area; biggestIndex = i }
    }
    let landmarks = result.faceLandmarks[biggestIndex]
    let box = Self.boundingBox(landmarks)

    // Roll: the 2D angle between the two outer-eye-corner landmarks — a
    // simple, well-understood atan2 on two known points, independent of
    // MediaPipe's transformation-matrix output entirely. Landmark indices
    // 33 (right eye outer corner) and 263 (left eye outer corner) are
    // MediaPipe Face Mesh's own STABLE, documented canonical indices (part
    // of the published 478-point topology, not something guessed at here).
    // This is the one angle in this module with real confidence behind it,
    // and the ONLY angle computation left in this file — see this file's
    // own header for why pitch/yaw's matrix decomposition was removed
    // rather than re-guessed.
    let rightEye = landmarks[33]
    let leftEye = landmarks[263]
    let rollRadians = atan2(Double(leftEye.y - rightEye.y), Double(leftEye.x - rightEye.x))
    let rollDegrees = rollRadians * 180 / .pi

    // Pitch/yaw: see this file's own header — the facialTransformationMatrixes
    // decomposition that used to be here was removed after a confirmed
    // build failure rather than re-guessed a second time blind. 0 reads as
    // "straight," the same honest fallback this already used whenever the
    // matrix was unavailable — not a regression from a working state, since
    // this decomposition never actually compiled, let alone ran.
    let pitchDegrees: Double = 0
    let yawDegrees: Double = 0

    let face = DetectedFace(
      bounds: DetectedFaceBounds(
        x: Double(box.minX), y: Double(box.minY),
        width: Double(box.width), height: Double(box.height)
      ),
      pitchAngle: pitchDegrees,
      rollAngle: rollDegrees,
      yawAngle: yawDegrees
    )
    stateLock.lock()
    lastResult = [face]
    stateLock.unlock()
  }

  // MARK: - Private helpers

  private static func boundingBox(_ landmarks: [NormalizedLandmark]) -> CGRect {
    var minX: CGFloat = 1, minY: CGFloat = 1, maxX: CGFloat = 0, maxY: CGFloat = 0
    for point in landmarks {
      minX = min(minX, point.x); maxX = max(maxX, point.x)
      minY = min(minY, point.y); maxY = max(maxY, point.y)
    }
    return CGRect(x: minX, y: minY, width: maxX - minX, height: maxY - minY)
  }

  /// Same VC v5 frame-orientation-string → UIImage.Orientation mapping the
  /// OCR-plus reference's HybridTextRecognizer.swift already uses (see this
  /// project's own research citing that file) — reused verbatim rather than
  /// re-derived, since it's a real, working, recently-verified mapping for
  /// this exact library version.
  private static func imageOrientation(from string: String) -> UIImage.Orientation {
    switch string {
    case "up": return .up
    case "down": return .down
    case "left": return .right
    case "right": return .left
    default: return .up
    }
  }
}

// Small NSObject-based forwarding proxy — FaceLandmarkerLiveStreamDelegate
// is an Objective-C-bridged protocol requiring NSObjectProtocol conformance,
// which HybridGlowFaceLandmarker itself can't provide (it inherits
// Nitrogen-generated HybridGlowFaceLandmarkerSpec_base, a plain Swift class
// — see this file's own header, error #1, confirmed by a real build
// failure, not a preemptive guess). This is the standard fix for that exact
// situation: a minimal NSObject subclass that does nothing but forward the
// one delegate callback back to its owner.
private class FaceLandmarkerDelegateProxy: NSObject, FaceLandmarkerLiveStreamDelegate {
  private weak var owner: HybridGlowFaceLandmarker?

  init(owner: HybridGlowFaceLandmarker) {
    self.owner = owner
    super.init()
  }

  func faceLandmarker(
    _ faceLandmarker: FaceLandmarker,
    didFinishDetection result: FaceLandmarkerResult?,
    timestampInMilliseconds: Int,
    error: Error?
  ) {
    owner?.handleResult(result, timestampInMilliseconds: timestampInMilliseconds, error: error)
  }
}

File: mobile/modules/mediapipe-face-landmarker/android/src/main/java/com/margelo/nitro/glowfacelandmarker/HybridGlowFaceLandmarker.kt
package com.margelo.nitro.glowfacelandmarker

import android.graphics.Bitmap
import android.util.Log
import androidx.annotation.Keep
import com.facebook.proguard.annotations.DoNotStrip
import com.google.mediapipe.framework.image.BitmapImageBuilder
import com.google.mediapipe.framework.image.MPImage
import com.google.mediapipe.tasks.core.BaseOptions
import com.google.mediapipe.tasks.vision.core.RunningMode
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarker
import com.google.mediapipe.tasks.vision.facelandmarker.FaceLandmarkerResult
import java.util.concurrent.atomic.AtomicBoolean

/**
 * Android Nitro HybridObject wrapping MediaPipe's Face Landmarker Tasks
 * Vision SDK as a live vision-camera frame processor — the Android
 * counterpart to HybridGlowFaceLandmarker.swift (iOS). Output reduced to
 * the exact same DetectedFace shape for the exact same reason: so
 * SkinScanCamera.tsx's existing gating math needs zero changes regardless
 * of which detector produced it — see GlowFaceLandmarker.nitro.ts's own
 * header comment.
 *
 * GROUNDING for the native-buffer piece specifically (the part built this
 * round, with extra care per explicit instruction): the
 * nativeCreateBitmapFromHardwareBuffer external fun below is implemented in
 * HybridGlowFaceLandmarker.cpp, itself a close adaptation of
 * react-native-vision-camera-ocr-plus's own real, working
 * HybridTextRecognizer.cpp — see that file's own header for the full
 * grounding, caveats, and the one deliberate defensive change made here
 * (always materializing a plain software bitmap, never handing a
 * HARDWARE-config one across the JNI boundary).
 *
 * NOT verified against a real device, a real Kotlin/NDK/Gradle build, or
 * even a syntax check — this environment has none of those (see this
 * project's own scope report). Written against MediaPipe's documented
 * Android Tasks Vision API surface. The facial transformation matrix
 * decomposition below carries the SAME flagged uncertainty as its iOS
 * counterpart (see that comment) — not verified against the real Android
 * API's exact matrix type/convention.
 */
@DoNotStrip
@Keep
class HybridGlowFaceLandmarker : HybridGlowFaceLandmarkerSpec() {

  // Implemented in HybridGlowFaceLandmarker.cpp — see this class's own
  // header for the grounding. Returns null on any native-side failure
  // (buffer lock failed, GPU path unavailable below API 31, etc.) —
  // detectFrame below treats that identically to "no face this frame,"
  // same as every other soft-failure path in this module.
  @DoNotStrip
  @Keep
  private external fun nativeCreateBitmapFromHardwareBuffer(pointer: Long): Bitmap?

  companion object {
    init {
      // Matches CMakeLists.txt's PACKAGE_NAME / add_library(...) name —
      // already loaded once via NitroGlowFaceLandmarkerOnLoad in the
      // ordinary autolinking path, this is the same defensive fallback
      // pattern the OCR-plus reference uses for its own equivalent load.
      try { System.loadLibrary("NitroGlowFaceLandmarker") } catch (_: UnsatisfiedLinkError) {}
    }
    private const val TAG = "GlowFaceLandmarker"
    // MediaPipe Face Mesh's own stable, documented canonical landmark
    // indices for the outer eye corners — identical topology/ordering on
    // iOS and Android (same underlying model), so this is the exact same,
    // real justification HybridGlowFaceLandmarker.swift's own roll
    // computation already relies on, not re-derived or guessed separately
    // here.
    private const val RIGHT_EYE_OUTER = 33
    private const val LEFT_EYE_OUTER = 263
  }

  private var landmarker: FaceLandmarker? = null
  private val isBusy = AtomicBoolean(false)
  @Volatile private var lastResult: Array<DetectedFace> = emptyArray()

  init {
    setupLandmarker()
  }

  private fun setupLandmarker() {
    val context = getApplicationContextOrNull()
    if (context == null) {
      Log.e(TAG, "No application context available — detection will return no faces")
      return
    }
    try {
      val baseOptions = BaseOptions.builder()
        // Loaded from this module's own bundled asset (see build.gradle's
        // own comment on how src/main/assets/face_landmarker.task merges
        // into the final app's assets) — resolved through the app's
        // AssetManager via this asset-relative path, not a filesystem path.
        .setModelAssetPath("face_landmarker.task")
        .build()
      val options = FaceLandmarker.FaceLandmarkerOptions.builder()
        .setBaseOptions(baseOptions)
        .setRunningMode(RunningMode.LIVE_STREAM)
        .setNumFaces(1) // only the primary/largest face — same as iOS
        // Same bug caught in self-review on the iOS side (see that file's
        // own comment) — defaults false; without it,
        // result.facialTransformationMatrixes() is always empty/absent and
        // pitch/yaw silently reads 0 forever.
        .setOutputFacialTransformationMatrixes(true)
        .setResultListener(this::onResult)
        .setErrorListener { error -> Log.e(TAG, "FaceLandmarker error: ${error.message}") }
        .build()
      landmarker = FaceLandmarker.createFromOptions(context, options)
    } catch (e: Exception) {
      // Same "fails to empty, never crashes the screen" contract as every
      // other native-piece-not-linked-yet case in this app (LightingSensor's
      // own ErrorBoundary is the JS-side equivalent of this same
      // philosophy) — detectFrame below simply returns emptyArray()
      // forever if this happens.
      Log.e(TAG, "FaceLandmarker failed to initialize: ${e.message}")
      landmarker = null
    }
  }

  // MARK: - HybridGlowFaceLandmarkerSpec

  override fun detectFrame(nativeBufferPointer: ULong, orientation: String): Array<DetectedFace> {
    val landmarker = this.landmarker ?: return emptyArray()

    // Never blocks the frame thread on MediaPipe inference — same
    // non-blocking isBusy/lastResult contract as the iOS side and the
    // OCR-plus reference's own scanFrame.
    if (isBusy.get()) return lastResult

    val bitmap = nativeCreateBitmapFromHardwareBuffer(nativeBufferPointer.toLong())
      ?: return lastResult // native conversion failed — see HybridGlowFaceLandmarker.cpp's own header for why this can legitimately happen (pixelFormat, API level)

    isBusy.set(true)
    try {
      val mpImage: MPImage = BitmapImageBuilder(bitmap).build()
      val timestampMs = System.currentTimeMillis()
      landmarker.detectAsync(mpImage, timestampMs)
      // detectAsync takes ownership of / retains what it needs from mpImage
      // internally (MediaPipe's own documented contract for LIVE_STREAM
      // mode) — the underlying bitmap is safe to let go of on this side
      // once this call returns; onResult below is where the actual
      // detection result — and any further reference to the frame's
      // pixel data — is handled.
    } catch (e: Exception) {
      isBusy.set(false)
      Log.e(TAG, "detectAsync submission failed: ${e.message}")
    }
    return lastResult
  }

  // MARK: - MediaPipe result listener

  private fun onResult(result: FaceLandmarkerResult, input: MPImage) {
    try {
      val allLandmarks = result.faceLandmarks()
      if (allLandmarks.isEmpty()) {
        lastResult = emptyArray()
        return
      }

      // Largest face wins — same convention as iOS.
      var biggestIndex = 0
      var biggestArea = -1f
      for (i in allLandmarks.indices) {
        val (w, h) = boundingSize(allLandmarks[i])
        val area = w * h
        if (area > biggestArea) { biggestArea = area; biggestIndex = i }
      }
      val landmarks = allLandmarks[biggestIndex]
      val bounds = boundingBox(landmarks)

      // Roll: same simple, well-understood 2D atan2 on two known eye
      // landmarks as iOS — the one angle in this module with real
      // confidence behind it (see this class's own header).
      val rightEye = landmarks[RIGHT_EYE_OUTER]
      val leftEye = landmarks[LEFT_EYE_OUTER]
      val rollRadians = Math.atan2(
        (leftEye.y() - rightEye.y()).toDouble(),
        (leftEye.x() - rightEye.x()).toDouble()
      )
      val rollDegrees = Math.toDegrees(rollRadians)

      // Pitch/yaw: same UNVERIFIED transformation-matrix decomposition
      // approach as iOS, with the same explicit low-confidence flag — see
      // HybridGlowFaceLandmarker.swift's own comment on this exact
      // tradeoff. Android's FaceLandmarkerResult.facialTransformationMatrixes()
      // returns Optional<MutableList<FloatArray>> per MediaPipe's
      // documented Android API — each FloatArray a 16-element row-major
      // 4x4 matrix. UNVERIFIED against the real installed library version;
      // wrapped defensively so a shape mismatch degrades to 0.0 (reads as
      // "straight") rather than an out-of-bounds crash or garbage value.
      var pitchDegrees = 0.0
      var yawDegrees = 0.0
      val matrices = result.facialTransformationMatrixes()
      if (matrices.isPresent && matrices.get().isNotEmpty()) {
        val m = matrices.get()[0]
        if (m.size >= 16) {
          // Row-major 4x4: m[8..11] is the third row (Z basis vector) in
          // this indexing convention — same standard R = Rz*Ry*Rx
          // extraction as iOS, same convention-uncertainty caveat.
          val m20 = m[8]; val m21 = m[9]; val m22 = m[10]
          pitchDegrees = Math.toDegrees(Math.atan2((-m21).toDouble(), Math.sqrt((m20 * m20 + m22 * m22).toDouble())))
          yawDegrees = Math.toDegrees(Math.atan2(m20.toDouble(), m22.toDouble()))
        }
      }

      lastResult = arrayOf(
        DetectedFace(
          bounds = bounds,
          pitchAngle = pitchDegrees,
          rollAngle = rollDegrees,
          yawAngle = yawDegrees
        )
      )
    } catch (e: Exception) {
      Log.e(TAG, "onResult processing failed: ${e.message}")
      lastResult = emptyArray()
    } finally {
      isBusy.set(false)
    }
  }

  // Copied verbatim (structure and reasoning) from the OCR-plus reference's
  // own getApplicationContextOrNull — Nitro HybridObjects have no
  // constructor-injected Context the way an Expo Module does (see this
  // module's own skin-segmentation sibling, which gets one via
  // appContext?.reactContext), so this reflects into
  // android.app.ActivityThread.currentApplication() the same way that
  // reference does for the identical problem. Grounded in real, working
  // code, not invented here.
  private fun getApplicationContextOrNull(): android.content.Context? {
    return try {
      val activityThread = Class.forName("android.app.ActivityThread")
      val currentApplicationMethod = activityThread.getMethod("currentApplication")
      currentApplicationMethod.invoke(null) as? android.app.Application
    } catch (_: Exception) {
      null
    }
  }

  private fun boundingSize(landmarks: List<com.google.mediapipe.tasks.components.containers.NormalizedLandmark>): Pair<Float, Float> {
    var minX = 1f; var minY = 1f; var maxX = 0f; var maxY = 0f
    for (p in landmarks) {
      if (p.x() < minX) minX = p.x()
      if (p.x() > maxX) maxX = p.x()
      if (p.y() < minY) minY = p.y()
      if (p.y() > maxY) maxY = p.y()
    }
    return Pair(maxX - minX, maxY - minY)
  }

  private fun boundingBox(landmarks: List<com.google.mediapipe.tasks.components.containers.NormalizedLandmark>): DetectedFaceBounds {
    var minX = 1f; var minY = 1f; var maxX = 0f; var maxY = 0f
    for (p in landmarks) {
      if (p.x() < minX) minX = p.x()
      if (p.x() > maxX) maxX = p.x()
      if (p.y() < minY) minY = p.y()
      if (p.y() > maxY) maxY = p.y()
    }
    return DetectedFaceBounds(
      x = minX.toDouble(), y = minY.toDouble(),
      width = (maxX - minX).toDouble(), height = (maxY - minY).toDouble()
    )
  }
}
